src/main/java/io/cdap/plugin/hubspot/common/HubspotHelper.java

```diff
@@ -0,0 +1,348 @@
+   * @param offset the offset is string type
+   * @return the instance of HubspotPage
+   * @throws IOException on issues with data reading
+   */
+  public HubspotPage getHubspotPage(SourceHubspotConfig sourceHubspotConfig, String offset) throws IOException {
+    HttpGet request = getRequest(sourceHubspotConfig, offset);
+
+    CloseableHttpResponse response = downloadPage(sourceHubspotConfig, request);
+    HttpEntity entity = response.getEntity();
+    String result;
+    if (entity != null) {
+      result = EntityUtils.toString(entity);
+      return parseJson(sourceHubspotConfig, result);
+    }
+    return null;
+  }
+
+  private CloseableHttpResponse downloadPage(SourceHubspotConfig sourceHubspotConfig,
+                                             HttpGet request) throws IOException {
+    HttpClientBuilder httpClientBuilder = HttpClientBuilder.create();
+    CloseableHttpClient client = httpClientBuilder.build();
+
+    int count = 0;
+    while (true) {
+      CloseableHttpResponse response = client.execute(request);
+      StatusLine statusLine = response.getStatusLine();
+      if (statusLine.getStatusCode() >= 300) {
+        if (500 > statusLine.getStatusCode() &&
+          statusLine.getStatusCode() >= 400) {
+          throw new IOException(statusLine.getReasonPhrase());
+        }
+        if (++count == MAX_TRIES) {
+          throw new IOException(statusLine.getReasonPhrase());
+        }
+      }
+      return response;
+    }
+  }
+
+  private HttpGet getRequest(SourceHubspotConfig sourceHubspotConfig, String offset) {
+    URI uri = null;
+    try {
+      URIBuilder b = new URIBuilder(getEndpoint(sourceHubspotConfig));
+      if (sourceHubspotConfig.apiKey != null) {
+        b.addParameter("hapikey", sourceHubspotConfig.apiKey);
+      }
+      if (sourceHubspotConfig.startDate != null) {
+        b.addParameter("start", sourceHubspotConfig.startDate);
+      }
+      if (sourceHubspotConfig.endDate != null) {
+        b.addParameter("end", sourceHubspotConfig.endDate);
+      }
+      for (String filter : sourceHubspotConfig.getFilters()) {
+        b.addParameter("f", filter);
+      }
+      if (getLimitPropertyName(sourceHubspotConfig) != null) {
+        b.addParameter(getLimitPropertyName(sourceHubspotConfig), PAGE_SIZE);
+      }
+      if (offset != null && getOffsetPropertyName(sourceHubspotConfig) != null) {
+        b.addParameter(getOffsetPropertyName(sourceHubspotConfig), offset);
+      }
+      uri = b.build();
+    } catch (URISyntaxException e) {
+      throw new RuntimeException("Can't build valid uri", e);
+    }
+    return new HttpGet(uri);
+  }
+
+  private HubspotPage parseJson(SourceHubspotConfig sourceHubspotConfig, String json) throws IOException {
+    JsonElement root = new JsonParser().parse(json);
+    JsonObject jsonObject = root.getAsJsonObject();
+    List<JsonElement> hubspotObjects = new ArrayList<>();
+    if (getObjectApiName(sourceHubspotConfig) != null) {
```

Comment:
`getObjectApiName()` is called in 3 places in this function. You can the object api name in a variable.
`String objectApiName = getObjectApiName(sourceHubspotConfig);`

Reply:
Done